typing/env.ml

```diff
@@ -2801,48 +2829,49 @@ end) = struct
         let map, shape = proj_shape map mod_shape (Shape.Item.class_type id) in
         map, add_cltype ?shape id decl env
 
-  let rec add_signature map mod_shape sg env =
+  let rec add_signature map mod_shape sg ?(mode = Mode.Value.(allow_right max)) env =
     match sg with
         [] -> map, env
     | comp :: rem ->
-        let map, env = add_item map mod_shape comp env in
-        add_signature map mod_shape rem env
+        let map, env = add_item map mod_shape comp mode env in
+        add_signature map mod_shape rem ~mode env
 end
 
-let add_signature =
+let add_signature map mod_shape sg ?mode env =
   let module M = Add_signature(Types)(struct
     let add_value ?shape ~mode id vd =
       add_value_lazy ?shape ~mode id (Subst.Lazy.of_value_description vd)
     let add_module_declaration = add_module_declaration
     let add_modtype = add_modtype
   end)
   in
-  M.add_signature
+  M.add_signature map mod_shape sg ?mode env
 
 let add_signature_lazy =
   let module M = Add_signature(Subst.Lazy)(struct
     let add_value ?shape ~mode = add_value_lazy ?check:None ?shape ~mode
-    let add_module_declaration =
-      add_module_declaration_lazy ~update_summary:true
+    let add_module_declaration ?arg ?shape ~check id pres md ?mode ?locks env =
+      add_module_declaration_lazy ~update_summary:true ?arg ?shape ~check id
+        pres md ?mode ?locks env
```

Comment:
Is this necessary? It seems like a very good use of partial evaluation and worth keeping

Reply:
Oof, I see now that it's a victim of the value restriction :-(